drivers/flash/flash_stm32u5x.c

```diff
@@ -0,0 +1,380 @@
+	 * that starts a cache invalidation procedure
+	 */
+	CLEAR_BIT(ICACHE->FCR, ICACHE_FCR_CBSYENDF);
+
+	LL_ICACHE_Disable();
+
+	/* Get tick */
+	tickstart = k_uptime_get_32();
+
+	/* Wait for instruction cache to get disabled */
+	while (LL_ICACHE_IsEnabled()) {
+		if ((k_uptime_get_32() - tickstart) >
+						ICACHE_DISABLE_TIMEOUT_VALUE) {
+			/* New check to avoid false timeout detection in case
+			 * of preemption.
+			 */
+			if (LL_ICACHE_IsEnabled()) {
+				status = -ETIMEDOUT;
+				break;
+			}
+		}
+	}
+
+	return status;
+}
+
+static void stm32u5x_icache_enable(void)
+{
+	LOG_DBG("I-cache Enable");
+	LL_ICACHE_Enable();
+}
+
+static int icache_wait_for_invalidate_complete(void)
+{
+	int status = -EIO;
+	uint32_t tickstart;
+
+	/* Check if ongoing invalidation operation */
+	if (LL_ICACHE_IsActiveFlag_BUSY()) {
+		/* Get tick */
+		tickstart = k_uptime_get_32();
+
+		/* Wait for end of cache invalidation */
+		while (!LL_ICACHE_IsActiveFlag_BSYEND()) {
+			if ((k_uptime_get_32() - tickstart) >
+					ICACHE_INVALIDATE_TIMEOUT_VALUE) {
+				break;
+			}
+		}
+	}
+
+	/* Clear any pending flags */
+	if (LL_ICACHE_IsActiveFlag_BSYEND()) {
+		LOG_DBG("I-cache Invalidation complete");
+
+		LL_ICACHE_ClearFlag_BSYEND();
+		status = 0;
+	} else {
+		LOG_ERR("I-cache Invalidation timeout");
+
+		status = -ETIMEDOUT;
+	}
+
+	if (LL_ICACHE_IsActiveFlag_ERR()) {
+		LOG_ERR("I-cache error");
+
+		LL_ICACHE_ClearFlag_ERR();
+		status = -EIO;
+	}
+
+	return status;
+}
+
+/*
+ * offset and len must be aligned on 8 for write,
+ * positive and not beyond end of flash
+ */
+bool flash_stm32_valid_range(const struct device *dev, off_t offset,
+			     uint32_t len,
+			     bool write)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+
+	if (((regs->OPTR & FLASH_OPTR_DBANK) == FLASH_OPTR_DBANK) &&
+			(CONFIG_FLASH_SIZE < STM32U5_SERIES_MAX_FLASH)) {
+		/*
+		 * In case of bank1/2 discontinuity, the range should not
+		 * start before bank2 and end beyond bank1 at the same time.
+		 * Locations beyond bank2 are caught by
+		 * flash_stm32_range_exists.
+		 */
+		if ((offset < BANK2_OFFSET) &&
+					(offset + len > FLASH_SIZE / 2)) {
+			return 0;
+		}
+	}
+
+	return (!write || (offset % 8 == 0 && len % 8 == 0U)) &&
+		flash_stm32_range_exists(dev, offset, len);
+}
+
+static int write_dword(const struct device *dev, off_t offset, uint64_t val)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	volatile uint32_t *flash = (uint32_t *)(offset
+						+ CONFIG_FLASH_BASE_ADDRESS);
+	uint32_t tmp;
+	int rc;
+
+	/* if the non-secure control register is locked,do not fail silently */
+	if (regs->NSCR & FLASH_NSCR_NSLOCK) {
+		LOG_ERR("NSCR locked\n");
+		return -EIO;
+	}
+
+	/* Check that no Flash main memory operation is ongoing */
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* Check if this double word is erased */
+	if ((flash[0] != 0xFFFFFFFFUL) || (flash[1] != 0xFFFFFFFFUL)) {
```

Comment:
just a note: a write will still fail if the ECC is already written.
ECC can be set if a programmer/ previous write was done with 0xFF...FF as data.
Maybe good to add this in the check?